pkg/assets/create/creater.go

```diff
@@ -0,0 +1,232 @@
+	// Multiple filters can be specified, in that case only files matching all filters will be returned.
+	Filters []assets.FileInfoPredicate
+
+	// Verbose if true will print out extra messages for debugging
+	Verbose bool
+
+	// StdErr allows to override the standard error output for printing verbose messages.
+	// If not set, os.StdErr is used.
+	StdErr io.Writer
+}
+
+// EnsureManifestsCreated ensures that all resource manifests from the specified directory are created.
+// This function will keep retrying creation until no errors are reported.
+// Pass the context to indicate how much time you are willing to wait until all resources are created.
+func EnsureManifestsCreated(ctx context.Context, manifestDir string, restConfig *rest.Config, options CreateOptions) error {
+	client, dc, err := getClientsFn(restConfig)
+	if err != nil {
+		return err
+	}
+
+	manifests, err := load(manifestDir, options)
+	if err != nil {
+		return err
+	}
+
+	if options.Verbose && options.StdErr == nil {
+		options.StdErr = os.Stderr
+	}
+
+	// Default QPS in client (when not specified) is 5 requests/per second
+	// This specify the interval between "create-all-resources", no need to make this configurable.
+	interval := 200 * time.Millisecond
+
+	// Retry creation until no errors are returned or the timeout is hit.
+	var (
+		lastCreateError      error
+		retryCount           int
+		mapper               meta.RESTMapper
+		needDiscoveryRefresh bool = true
+	)
+	err = wait.PollImmediateUntil(interval, func() (bool, error) {
+		retryCount++
+		// If we get rest mapper error, we need to pull updated discovery info from API server
+		if needDiscoveryRefresh {
+			mapper, err = fetchLatestDiscoveryInfoFn(dc)
+			if err != nil {
+				return false, err
+			}
+		}
+		err, needDiscoveryRefresh = create(manifests, client, mapper, options)
+		if err == nil {
+			// No errors means all resources were created successfully, reset lastCreateError and exit the loop.
+			lastCreateError = nil
+			return true, nil
+		}
+		if options.Verbose {
+
+		}
+		lastCreateError = err
+		if options.Verbose {
+			fmt.Fprintf(options.StdErr, "[#%d] %s\n", retryCount, lastCreateError)
+		}
+		return false, nil
+	}, ctx.Done())
+
+	// Return the last observed set of errors from the create process instead of timeout error.
+	if lastCreateError != nil {
+		return lastCreateError
+	}
+
+	return err
+}
+
+// allow to override in unit test
+var getClientsFn = getClients
+
+func getClients(config *rest.Config) (dynamic.Interface, *discovery.DiscoveryClient, error) {
+	client, err := dynamic.NewForConfig(config)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	dc, err := discovery.NewDiscoveryClientForConfig(config)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return client, dc, nil
+}
+
+// allow to override in unit test
+var fetchLatestDiscoveryInfoFn = fetchLatestDiscoveryInfo
+
+func fetchLatestDiscoveryInfo(dc *discovery.DiscoveryClient) (meta.RESTMapper, error) {
```

Comment:
there is `CachedDiscoveryInterface` with an `Invalidate()` func, created via `NewMemCacheClient`.

Reply:
settled via IRC, added TODO as this does not have to be elegant on first iteration. Also CachedDiscoveryInterface might complicate the unit test and it might not be worth the benefits.